File: algorithms/cpp/673.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int findNumberOfLIS(vector<int> &nums) {
        int n = nums.size();
        if (n == 0) return 0;
        vector<int> maxLength(n, 1), count(n, 1);
        for (int i = 0; i < n; i++) {
            for (int j = i+1; j < n; j++) {
                if (nums[i] >= nums[j]) continue;
                if (maxLength[i]+1 > maxLength[j]) {
                    maxLength[j] = maxLength[i] + 1;
                    count[j] = count[i];
                }
                else if (maxLength[i]+1 == maxLength[j])
                    count[j] += count[i];
            }
        }
        int length = *max_element(maxLength.begin(), maxLength.end());
        int result = 0;
        for (int i = 0; i < n; i++)
            if (maxLength[i] == length)
                result += count[i];
        return result;
    }
};

int main() {
    return 0;
}

File: algorithms/python/673.py
class Solution(object):
    def findNumberOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums: return 0
        n = len(nums)
        dp, cnt = [1]*n, [1]*n
        for i in range(n):
            for j in range(i+1, n):
                if nums[i] >= nums[j]:
                    continue
                if dp[i]+1 > dp[j]:
                    dp[j] = dp[i]+1
                    cnt[j] = cnt[i]
                elif dp[i]+1 == dp[j]:
                    cnt[j] += cnt[i]
        maxLen = max(dp)
        rls = sum(map(lambda x: cnt[x[0]] if dp[x[0]] == maxLen else 0, enumerate(dp)))
        return rls